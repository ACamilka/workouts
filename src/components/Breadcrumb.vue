<template>
  <el-breadcrumb separator="/">
    <el-breadcrumb-item v-for="(breadcrumb, idx) in breadcrumbList"
                        :to="{ name: breadcrumb.name }"
                        :key="idx">
      {{ breadcrumb.title }}
    </el-breadcrumb-item>
  </el-breadcrumb>
</template>

<script>

  import { mapGetters } from 'vuex';

  export default {
    name: "Breadcrumb",
    data() {
      return {
        breadcrumbList: []
      }
    },
    mounted() {
      this.update();
    },
    computed: {
      ...mapGetters({ currentWorkout: 'workout' })
    },
    watch: {
      '$route': 'update',
      'currentWorkout': 'update'
    },
    methods: {
      update() {
        const route = this.$route;
        this.breadcrumbList = route.matched.reduce((dest, src) => {
          const meta = src.meta;
          if (meta && meta.breadcrumb && meta.breadcrumb.title) {
            let title = meta.breadcrumb.title;
            if (title instanceof Function) {
              title = title(this.currentWorkout)
            }
            dest.push({
              title: title,
              name: src.name,
            });
          }
          return dest;
        }, []);
      }
    }
  }
</script>

<style scoped>

</style>