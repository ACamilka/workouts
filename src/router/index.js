import Vue from 'vue';
import VueRouter from 'vue-router';

import Container from "@/components/Container";
import WorkoutList from "@/views/WorkoutList";
import WorkoutDetails from "@/views/WorkoutDetails";

Vue.use(VueRouter);

const routes = [
  { path: '/',
    name: 'container',
    component: Container,
    redirect: '/workouts',
    meta: { breadcrumb: { title: 'Workouts' } },
    children: [
      { path: 'workouts',
        name: 'workouts',
        component: WorkoutList },
      { path: 'workouts/:id',
        name: 'workout',
        component: WorkoutDetails,
        meta: { breadcrumb: { title: ({ name }) => `${name}` } } },
    ]
  }
];

const router = new VueRouter({
  routes
});

export default router;