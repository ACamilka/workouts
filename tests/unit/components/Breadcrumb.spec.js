import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from 'vuex';
import ElementUI from "element-ui";

import router from "@/router";
import Breadcrumb from "@/components/Breadcrumb";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(ElementUI);

describe('Breadcrumb', () => {

  let getters;
  let store;
  const workout = {
    id: "a461e9bc-6fa4-4eae-be0d-c52c1fc7fa43",
    name: 'test workout'
  };

  beforeEach(() => {
    getters = {
      workout: () => workout
    };
    store = new Vuex.Store({ getters });
  });

  it('should display workout details breadcrumb', () => {
    router.push({ name: 'workout', params: { id: workout.id }});
    const wrapper = shallowMount(Breadcrumb, { store, localVue, router, });
    expect(wrapper.vm.breadcrumbList.length).toBe(2);
    expect(wrapper.vm.breadcrumbList).toMatchSnapshot();
  });

});