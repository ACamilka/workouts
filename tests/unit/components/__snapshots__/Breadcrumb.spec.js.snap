// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb should display workout details breadcrumb 1`] = `
Array [
  Object {
    "name": "container",
    "title": "Workouts",
  },
  Object {
    "name": "workout",
    "title": "test workout",
  },
]
`;
